plugins/backstage-plugin-coder/src/api/Auth.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file Defines shared values and types used among any custom Coder auth
+ * implementations for the frontend.
+ */
+import { createApiRef } from '@backstage/core-plugin-api';
+import { CODER_API_REF_ID_PREFIX } from '../typesConstants';
+
+/**
+ * Data about the auth that can safely be exposed throughout multiple parts of
+ * the application without too much worry about security.
+ *
+ * All values should be JSON-serializable.
+ */
+export type SafeAuthData = Readonly<{
+  isTokenValid: boolean;
+  tokenHash: number | null;
+  initialTokenHash: number | null;
+  isInsideGracePeriod: boolean;
+}>;
+
+export type AuthSubscriptionCallback = (payload: SafeAuthData) => void;
+export type AuthValidatorDispatch = (newStatus: boolean) => void;
+
+/**
+ * Shared set of properties among all Coder auth implementations
+ */
+export type CoderAuthApi = SafeAuthData & {
+  /**
+   * Requests the token from the auth class. This may not always succeed (e.g.,
+   * auth doesn't think it's safe, there is no token to provide), in which case,
+   * the value will be null.
+   */
+  requestToken: () => string | null;
+
+  /**
+   * Gives back a "state setter" that lets a different class dispatch a new auth
+   * status to the auth class implementation.
+   *
+   * Use this to send the new status you think the auth should have. The auth
+   * will decide whether it will let the dispatch go through and update state.
+   */
+  getAuthStateSetter: () => AuthValidatorDispatch;
+
+  /**
+   * Subscribes an external system to auth changes.
+   *
+   * Returns an pre-wired unsubscribe callback to remove fuss of needing to hold
+   * onto the original callback if it's not directly needed anymore
+   */
+  subscribe: (callback: AuthSubscriptionCallback) => () => void;
```

Comment:
So, I wish I could use Context more, but the main issue is that contexts aren't really the "Backstage way", and a lot of their APIs and tools are built on their own bespoke systems

So there were two things I was shooting for with the API classes (and I'm typing so much to make sure this is documented somewhere):
- For the end user, it centralizes the logic for setting up the plugin
   - When they add the plugin, they can also add the APIs that they want (the main example I'm thinking of is that we could have a single Coder auth API ref, and let the user choose whether they want to bring in a token auth implementation or an oauth implementation to be bound to it)
   - It could also help us get rid of the `CoderConfig` object over time. I feel like it's not clear where exactly to put it, other than sorta close-ish to where the CoderProvider is being used, and you have to figure out where you want to put it in the giant page of components. If this were an API class, there would be one single, obvious spot: where you add the plugin, and you wouldn't have to open any other files
- It becomes easier to add the API class logic to their helpers for spinning up plugins in isolation
   - The reason we can't do it right now is that when Backstage spins up, it only adds the dependencies it's been told it needs. For the full deployment, it adds everything. But when you spin up a plugin separately, Backstage will only add the APIs that have explicitly been registered with it. We currently don't have anything registered, so the plugin hangs
   - I was thinking that having our own API classes would be the easiest way to bring the Backstage API dependencies in, because again, you can do it all at the plugin definition

But maybe it doesn't have to be like that – maybe we could register the APIs from Backstage, and then handle everything else through React and `useApi`. It might not be as flexible, but it might be a lot easier to work with

Reply:
But yeah, complexity is definitely a worry for me, too, and even if the code gets merged in as-is, I think there will still be some complexity to shave off

I made [an Excalidraw drawing](https://excalidraw.com/#json=l5u4NcBdw-LbncpKM5Quo,kyp4bHYPQqmq3RQjITRRGg) to give a top-level view of how these things are supposed to fit together, but I feel like (1) the implementation isn't living up to how neat the drawing looks, and (2) maybe even some of the arrows linking things together don't need to be there

I haven't done much OOP professionally, but this is where I wish I had more Go experience, because I feel like I'd have no choice but to figure out a solution that doesn't get too clever, and that's easier to maintain

My gut feeling is that I want these interfaces to be as small as humanly possible (maybe even one method/value total), but I don't know what the most effective way of doing that is

Reply:
Helpful chart!  The picture does become clearer with the goal of being able to slot auth implementations into the system, although I keep feeling there should be a more Backstage-oriented way, similar to how their built-in auth providers work.  I think it is not unreasonable to only support one hard-coded method either, unless we have business requirements saying otherwise.

Thinking about it more, and stealing from Roadie's [pull request plugin](https://github.com/RoadieHQ/roadie-backstage-plugins/blob/fb8ce775546faa2fb68fe922d946e1ab5909aeb2/plugins/frontend/backstage-plugin-github-pull-requests/src/components/useGithubRepositoryData.tsx#L29), is this what Backstage means for it look like in the end?

```
export const useCoderClient = (url: string) => {
  const coderAuthApi = useApi(coderAuthApiRef);
  return useAsync(async (): Promise<CoderClient> => {
    const token = await coderAuthApi.getAccessToken(['*']);
    return new Client(url, token));
  }, [url, coderAuthApi]);
};
```

Where the Coder auth API implements a Backstage auth provider.  Although, not sure we can actually implement the current localstorage-based auth as an auth provider; the OAuth-based one should be fine though.  I suppose it could be like this to start, really one could do anything as long as it spits out a token in the end:

```
export const useCoderClient = (url: string, token: String) => {
  return useAsync(async (): Promise<CoderClient> => {
    return new Client(url, token));
  }, [url, token]);
};
```

My sense, from reading the code and the chart, is that much of this is mainly for tying the client and auth together (for the client to subscribe to changes to auth), and I think maybe they should actually be tied in a more top-down way, whether that is a context or a hook, which would simplify a few things.

All this said, I might be completely missing the mark still haha, I think I need to actually get in there to really understand; looking forward to tomorrow!

Reply:
I think this file change also shows what I was hoping to achieve: [old](https://github.com/coder/backstage-plugins/blob/main/plugins/backstage-plugin-coder/src/hooks/useCoderWorkspacesQuery.ts) vs [new](https://github.com/coder/backstage-plugins/blob/mes/api-factories-04/plugins/backstage-plugin-coder/src/hooks/useCoderWorkspacesQuery.ts)

My hope was that instead of having to manually pass every single dependency into every API function call, they would all be co-located into one single, thick client.

I think I achieve that for this file – it's just, everything else around it has become a lot more chaotic

Reply:
Just for posterity's sake, we talked about it, the part I was missing is that we want to end up with something like:

```
createApiFactory(authApiRef, () => new CoderAuthApi())
createApiFactory(coderRef, (authApi) => new CoderApi(authApi))
```

1. This means `CoderApi` needs to have the ability to subscribe to changes in `CoderAuthApi` for when the token updates, as Backstage will not recreate `CoderApi` if `authApi` changes and does not seem to have any built-in mechanism for updating an API reference with changes from another API reference.
2. We also need subscriptions to use the token status in a component, since if you try to use an API reference in a dependency array, it will **not** trigger any updates.

We could do it all React-based instead with hooks or contexts or whatever, but it is not the Backstage way with API factories.  And, the OAuth provider **has** to be an API factory, so at the very least we are stuck with auth being an API reference, regardless of what we do with the client, which means because of 2 we still need this code.